platform/golden/provisioners/roles/override_kernel_parameters/files/dropin-profile-kvm-for-grub.cfg

```diff
@@ -0,0 +1,2 @@
+GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX modprobe.blacklist=nouveau pci=realloc=on pcie_ports=native usbcore.nousb vsyscall=none intel_iommu=on,sm_on amd_iommu=on amd_iommu_intr=vapic iommu=pt iommu.strict=1 iommu.forcedac=1 kvm-amd.avic=1 kvm-amd.nested=0 kvm-amd.npt=1 transparent_hugepage=never nmi_watchdog=0 default_hugepagesz=2M hugepagesz=2M hugepages=400000 video=efifb:off iommu.passthrough=1 iommu.forcedac=1 rd.driver.pre=vfio-pci pcie_port_pm=off pcie_bus_perf pcie_aspm=off"
```

Comment:
Some of these are...surprising, I guess? Not wrong, not bad, but I would hope that there's a good reason for them, even if that reason is "doesn't boot unless these are set."

One surprise is that `nouveau` must be blacklisted. I know that we don't want it loaded, but it should automatically unload via the VGA arbiter as nVidia's driver is loaded. Similarly, explicitly disabling `efifb` is usually not what we want, since it makes debugging early boot much harder and should also be unloaded via VGA arbiter.

Another, more serious surprise is that huge pages need explicit configuration. What goes wrong if we allow them by default?

Reply:
Hello Corbin.I blacklist nouveau because, otherwise, kernel oops after kernel oops happens, until all processes are wedged. I do not understand the root cause in detail, however, I have very gooed information that indicates the problem is related to loading the nvidia kernel driver, or perhaps just compiling it.

efifb is some cargo copy cult I picked up. After a long time of keeping it on, I prefer it, but it does cut the early boot log.

If the hugepage count is not set, then nothing special happens. If the hugepage count is set, and there is memory to allocate a a DMA-safe as specified by 400000*2MB (=800GB of ram), then violat.

I do agree, the inability to set this with percentages is also surprising to me.

Perhaps kernel engineers are engineers who do not value a ratio.

Keep in mind, I read the irritating boot command line docus about 50 times.

If you have any suggestions for improvement, please to accept!

Reply:
@MostAwesomeDude I need to point out that this KVM kernel command line configuration is *only* for metal. I also have a configuration for virtual machines that run virtual machines (used by our builder). I may have a configuration, as well, for generalalized GPU workloads.

I have disabled this role for now, as the only image we create is a generalized GPU virtual machine.

Reply:
@MostAwesomeDude I think what you had asked about hugepages may be related to transparent huge pages. THP causes performance degradation and serious memory fragmentation. Worse, you are not guaranteed pages you request via the `hugetlbfs`. Requests become hints. Hard-coding the page count (under <9 0%, as the kernel has a hard limit on hard page allocation) results in better consistency and stability if huge pages are desired. I hope that helps.

I think this file should be broken into 3 parts - `amd`, `intel`, and `general`. I am not super keen to generate an image separately for `intel` and `amd` though, which is why I left it as is.

This is a work in progress.